### PR TITLE
Implement the decodeResponse / encodeArgument in the arduino end. 

### PR DESCRIPTION
**Pull request by hbgxsm95**
Manage to run it on the Raspberry Pi end and Arduino. 
Fix some minor bugs including missing comma in the encodeResponse in the textParser.
If there is anything wrong with the parse in the Arduino end, it would blink the LED and delay for 1 seconds. Then resend the data request.
The current system use TextParser in both ends.

### CODE REVIEW

File: go/iot/protocol/examples/arduino/protocol.ino [hunk 1]
```diff
@@ -134,8 +149,12 @@ void readLine(char * buffer, int maxLength)
   buffer[idx] = 0;
 }
 
+response* parseResponse(char *json){
```
Comment (songy23) [inline go/iot/protocol/examples/arduino/protocol.ino]: nit: s/json/response

File: go/iot/protocol/examples/arduino/protocol.ino [hunk 2]
```diff
@@ -134,8 +149,12 @@ void readLine(char * buffer, int maxLength)
...
+  //return parseResponseJson(json);
+  return parseResponseText(json);
```
Comment (songy23) [inline go/iot/protocol/examples/arduino/protocol.ino]: ditto

File: go/iot/protocol/examples/arduino/protocol.ino [hunk 3]
```diff
@@ -154,11 +173,65 @@ JsonObject* parseResponse(char *json) {
...
     return NULL;
   } else{
-    return &root;
+    response *res = new response(root["Code"], root["Info"]);
+    return res;
+  }
+}
+// Typical example is {"Code" : 200, "Info" : "TMP"}
+// There could be whitespace inside the message
+response* parseResponseText(char *json) {
```
Comment (songy23) [inline go/iot/protocol/examples/arduino/protocol.ino]: s/json/text

File: go/iot/protocol/examples/arduino/protocol.ino (hunk 1 above)
Comment (hbgxsm95) [reply go/iot/protocol/examples/arduino/protocol.ino]: got.

File: go/iot/protocol/opencensus/opencensus_base.go [hunk 4]
```diff
@@ -126,8 +126,10 @@ func (census *OpenCensusBase) Record(arguments *protocol.MeasureArgument) *proto
 			switch vv := measure.(type) {
 			case *stats.Float64Measure:
 				stats.Record(ctx, vv.M(float64(value)))
+				break
```
Comment (rakyll) [inline go/iot/protocol/opencensus/opencensus_base.go]: You don't need break. In Go, case by default breaks unlike C.

File: go/iot/protocol/examples/arduino/protocol.ino [hunk 5]
```diff
@@ -154,11 +172,65 @@ JsonObject* parseResponse(char *json) {
...
 }
+// Typical example is {"Code" : 200, "Info" : "TMP"}
+// There could be whitespace inside the message
+response* parseResponseText(char *text) {
+  String ss(text);
+  String codeKey = "\"Code\"";
+  String infoKey = "\"Info\"";
+  int codeKeyPos = ss.indexOf(codeKey);
+  int infoKeyPos = ss.lastIndexOf(infoKey);
+  if (codeKeyPos == -1 || infoKeyPos == -1)
```
Comment (rakyll) [inline go/iot/protocol/examples/arduino/protocol.ino]: Add curly braces.

if ... {
}

File: go/iot/protocol/examples/arduino/protocol.ino [hunk 6]
```diff
@@ -154,11 +172,65 @@ JsonObject* parseResponse(char *json) {
...
+    return NULL;
+  codeKeyPos = codeKeyPos + codeKey.length();
+  infoKeyPos = infoKeyPos + infoKey.length();
+  // Find the code value
+  // Value is between the ',' and ':'
+  int colonIndex = codeKeyPos;
+  while(colonIndex < ss.length() && ss[colonIndex] != ':') colonIndex++;
```
Comment (rakyll) [inline go/iot/protocol/examples/arduino/protocol.ino]: Add curly braces always. It avoids bugs in the long term.

File: go/iot/protocol/examples/arduino/protocol.ino [hunk 7]
```diff
@@ -154,11 +172,65 @@ JsonObject* parseResponse(char *json) {
...
+  if (colonIndex == ss.length()) return NULL;
```
Comment (rakyll) [inline go/iot/protocol/examples/arduino/protocol.ino]: Ditto

File: go/iot/protocol/examples/arduino/protocol.ino [hunk 8]
```diff
@@ -154,11 +172,65 @@ JsonObject* parseResponse(char *json) {
...
+  int commaIndex = colonIndex;
+  while(commaIndex < ss.length() && ss[commaIndex] != ',') commaIndex++;
```
Comment (rakyll) [inline go/iot/protocol/examples/arduino/protocol.ino]: Ditto

File: go/iot/protocol/examples/arduino/protocol.ino [hunk 9]
```diff
@@ -154,11 +172,65 @@ JsonObject* parseResponse(char *json) {
...
+  if (commaIndex == ss.length()) return NULL;
```
Comment (rakyll) [inline go/iot/protocol/examples/arduino/protocol.ino]: Ditto

File: go/iot/protocol/examples/arduino/protocol.ino [hunk 10]
```diff
@@ -154,11 +172,65 @@ JsonObject* parseResponse(char *json) {
...
+  String valuess = ss.substring(colonIndex + 1, commaIndex);
+  valuess.trim();
+  int value = toInteger(valuess);
+  if (value == -1) return NULL;
```
Comment (rakyll) [inline go/iot/protocol/examples/arduino/protocol.ino]: Ditto

File: go/iot/protocol/examples/arduino/protocol.ino [hunk 11]
```diff
@@ -154,11 +172,65 @@ JsonObject* parseResponse(char *json) {
...
+
+  //exclusive the right bracket
+  String info = ss.substring(infoKeyPos + 1, ss.length() - 1);
+  int leftMark = info.indexOf('\"');
+  int rightMark = info.lastIndexOf('\"');
+  if (leftMark == -1 || rightMark == -1 || leftMark == rightMark)
```
Comment (rakyll) [inline go/iot/protocol/examples/arduino/protocol.ino]: Curly.

File: go/iot/protocol/examples/arduino/protocol.ino [hunk 12]
```diff
@@ -154,11 +172,65 @@ JsonObject* parseResponse(char *json) {
...
+  info = info.substring(leftMark + 1, rightMark);
+
+  response *res = new response(value, info);
+  return res;
+}
+
+int toInteger(String code){
+  int res = 0;
+  for (int index = 0; index < code.length(); index++){
+    if (code[index] > '9' || code[index] < '0')
```
Comment (rakyll) [inline go/iot/protocol/examples/arduino/protocol.ino]: Curly

File: go/iot/protocol/examples/arduino/protocol.ino [hunk 13]
```diff
@@ -134,8 +148,12 @@ void readLine(char * buffer, int maxLength)
   buffer[idx] = 0;
 }
 
+response* parseResponse(char *response){
+  //return parseResponseJson(response);
```
Comment (rakyll) [inline go/iot/protocol/examples/arduino/protocol.ino]: You can remove these legacy lines. You can revert them back via the history if they are needed for reference.

File: go/iot/protocol/examples/arduino/protocol.ino [hunk 14]
```diff
@@ -154,11 +172,65 @@ JsonObject* parseResponse(char *json) {
...
+  int rightMark = info.lastIndexOf('\"');
+  if (leftMark == -1 || rightMark == -1 || leftMark == rightMark)
+    return NULL;
+  info = info.substring(leftMark + 1, rightMark);
+
+  response *res = new response(value, info);
+  return res;
+}
+
+int toInteger(String code){
```
Comment (rakyll) [inline go/iot/protocol/examples/arduino/protocol.ino]: Isn't toInt working for your case? https://www.arduino.cc/en/Tutorial.StringToIntExample

File: go/iot/protocol/examples/arduino/protocol.ino [hunk 15]
```diff
@@ -178,12 +250,11 @@ void sendData() {
...
   root["Name"] = "opencensus.io/measure/Temperature";
-  float temp = myHumidity.readTemperature();
-  root["Measurement"] = String(temp);
+  root["Value"] = String(temp);
 
-  JsonObject& tagPairs = root.createNestedObject("Tag");
+  JsonObject& tagPairs = root.createNestedObject("Tags");
   tagPairs["ArduinoId"] = "Arduino-1";
-  tagPairs["Date"] = "2018-07-02";
+  tagPairs["Date"] = "2018-08-02";
```
Comment (rakyll) [inline go/iot/protocol/examples/arduino/protocol.ino]: Any way you can programmatically generate this?

File: go/iot/protocol/examples/arduino/protocol.ino (hunk 14 above)
Comment (hbgxsm95) [reply go/iot/protocol/examples/arduino/protocol.ino]: toInt would return 0 for those invalid value.  And for string like "123s123" it would return 123.

**Pull request merged**